File: VB/MVVMDemo.Wpf/Modules/AsyncCommands/AsyncDelegateCommandsViewModel.vb
Imports DevExpress.Mvvm
Imports System
Imports System.Threading
Imports System.Threading.Tasks

Namespace MVVMDemo.AsyncCommands
    Public Class AsyncDelegateCommandsViewModel
        Inherits ViewModelBase

        Private privateCalculateCommand As AsyncCommand
        Public Property CalculateCommand() As AsyncCommand
            Get
                Return privateCalculateCommand
            End Get
            Private Set(ByVal value As AsyncCommand)
                privateCalculateCommand = value
            End Set
        End Property

        Private Function Calculate() As Task
            Return Task.Factory.StartNew(AddressOf CaclulateCore)
        End Function
        Private Sub CaclulateCore()
            For i As Integer = 0 To 100
                If CalculateCommand.IsCancellationRequested Then
                    Progress = 0
                    Return
                End If
                Progress = i
                Thread.Sleep(TimeSpan.FromMilliseconds(20))
            Next i
        End Sub

        Private _Progress As Integer
        Public Property Progress() As Integer
            Get
                Return _Progress
            End Get
            Set(ByVal value As Integer)
                SetProperty(_Progress, value, Function() Progress)
            End Set
        End Property

        Public Sub New()
            CalculateCommand = New AsyncCommand(AddressOf Calculate)
        End Sub

    End Class
End Namespace

File: CS/MVVMDemo.Wpf/Modules/AsyncCommands/AsyncDelegateCommandsViewModel.cs
using DevExpress.Mvvm;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace MVVMDemo.AsyncCommands {
    public class AsyncDelegateCommandsViewModel : ViewModelBase {
        public AsyncCommand CalculateCommand { get; private set; }

        Task Calculate() {
            return Task.Factory.StartNew(CaclulateCore);
        }
        void CaclulateCore() {
            for(int i = 0; i <= 100; i++) {
                if(CalculateCommand.IsCancellationRequested) {
                    Progress = 0;
                    return;
                }
                Progress = i;
                Thread.Sleep(TimeSpan.FromMilliseconds(20));
            }
        }

        int _Progress;
        public int Progress {
            get { return _Progress; }
            set { SetProperty(ref _Progress, value, () => Progress); }
        }

        public AsyncDelegateCommandsViewModel() {
            CalculateCommand = new AsyncCommand(Calculate);
        }

    }
}
